module/modmanager/manager.go

```diff
@@ -403,6 +404,78 @@ func (mgr *Manager) ValidateConfig(ctx context.Context, conf resource.Config) ([
 	return resp.Dependencies, nil
 }
 
+// AddImplicitDependencies modifies the config diff to add implicit dependencies to changed components
+// and updates modular components & services whose module has been changed with new implict deps and adds them to conf.Modified.
+func (mgr *Manager) AddImplicitDependencies(ctx context.Context, conf *config.Diff) error {
+	mgr.mu.RLock()
+	defer mgr.mu.RUnlock()
+	validateModularResource := func(conf *resource.Config) {
+		if mgr.Provides(*conf) {
+			implicitDeps, err := mgr.ValidateConfig(ctx, *conf)
+			if err != nil {
+				mgr.logger.Errorw("modular config validation error found in resource: "+conf.Name, "error", err)
+			}
+
+			// Modify resource to add its implicit dependencies.
+			conf.ImplicitDependsOn = implicitDeps
+		}
+	}
+	// Add implicit deps for all components in the new config
+	for _, c := range conf.Right.Components {
+		addedIndex := slices.IndexFunc(conf.Added.Components, func(elem resource.Config) bool { return elem.Name == c.Name })
+		if addedIndex >= 0 {
+			validateModularResource(&conf.Added.Components[addedIndex])
+			continue
+		}
+
+		modifiedIndex := slices.IndexFunc(conf.Modified.Components, func(elem resource.Config) bool { return elem.Name == c.Name })
+		if modifiedIndex >= 0 {
+			validateModularResource(&conf.Modified.Components[modifiedIndex])
+			continue
+		}
+
+		// See if this component is being provided by a module
+		mod, ok := mgr.getModule(c)
+		if !ok {
+			continue
+		}
+		// If it is, check against the modified modules to determine if the config should also be updated
+		if slices.ContainsFunc(conf.Modified.Modules, func(elem config.Module) bool { return elem.Name == mod.name }) {
+			validateModularResource(&c)
+			conf.Modified.Components = append(conf.Modified.Components, c)
+			continue
+		}
+	}
```

Comment:
I've butchered the somewhat elegant code that benji wrote and have turned it into a mess of index based logic and loop comparisons. This is not great. 

Unfortunately, config diffs do not currently have a `diff.Unchanged` field so I need to loop through all of the diff to find components that are unchanged (and along the way, I can update the added and modified resources)

Suggestions are welcome if there is a elegant solution I have missed